File: gravitee-am-ui/src/app/services/trust-domain.service.ts
/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { AppConfig } from '../../config/app.config';
import { normalizeTrustDomain } from '../domain/settings/trust-domains/trust-domain.types';

@Injectable()
export class TrustDomainService {
  private domainBaseURL = AppConfig.settings.domainBaseURL;

  constructor(private http: HttpClient) {}

  list(domainId: string): Observable<any> {
    return this.http
      .get<any>(this.domainBaseURL + domainId + '/trust-domains')
      .pipe(map((trustDomains) => (trustDomains ?? []).map(normalizeTrustDomain)));
  }

  get(domainId: string, id: string): Observable<any> {
    return this.http.get<any>(this.domainBaseURL + domainId + '/trust-domains/' + id).pipe(map(normalizeTrustDomain));
  }

  create(domainId: string, trustDomain: any): Observable<any> {
    return this.http.post<any>(this.domainBaseURL + domainId + '/trust-domains', trustDomain).pipe(map(normalizeTrustDomain));
  }

  update(domainId: string, id: string, trustDomain: any): Observable<any> {
    return this.http
      .put<any>(this.domainBaseURL + domainId + '/trust-domains/' + id, {
        name: trustDomain.name,
        description: trustDomain.description,
        spiffeTrustDomain: trustDomain.spiffeTrustDomain,
        issuer: trustDomain.issuer,
        keyMaterial: trustDomain.keyMaterial,
        refreshIntervalSeconds: trustDomain.refreshIntervalSeconds,
        allowedAlgorithms: trustDomain.allowedAlgorithms,
        scopeMappings: trustDomain.scopeMappings,
        userBindingEnabled: trustDomain.userBindingEnabled,
        userBindingCriteria: trustDomain.userBindingCriteria,
      })
      .pipe(map(normalizeTrustDomain));
  }

  delete(domainId: string, id: string): Observable<any> {
    return this.http.delete<any>(this.domainBaseURL + domainId + '/trust-domains/' + id);
  }
}

File: gravitee-am-service/src/main/java/io/gravitee/am/service/TrustDomainService.java
/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.am.service;

import io.gravitee.am.identityprovider.api.User;
import io.gravitee.am.model.Domain;
import io.gravitee.am.model.Reference;
import io.gravitee.am.model.ReferenceType;
import io.gravitee.am.model.oidc.CrossAppAccessResourceServerView;
import io.gravitee.am.model.oidc.TrustedDomain;
import io.gravitee.am.service.model.NewTrustDomain;
import io.gravitee.am.service.model.NewTrustedDomain;
import io.gravitee.am.service.model.UpdateTrustDomain;
import io.gravitee.am.service.model.UpdateTrustedDomain;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Single;

public interface TrustDomainService {

    Maybe<TrustedDomain> findById(String id);

    /**
     * Looks up a trusted domain by the label it is known by, unique within a reference.
     */
    Maybe<TrustedDomain> findByName(ReferenceType referenceType, String referenceId, String name);

    /**
     * Looks up the trusted domain that vouches for a SPIFFE trust domain, unique within a reference.
     */
    Maybe<TrustedDomain> findBySpiffeTrustDomain(ReferenceType referenceType, String referenceId, String spiffeTrustDomain);

    Flowable<TrustedDomain> findByReference(ReferenceType referenceType, String referenceId);

    Flowable<CrossAppAccessResourceServerView> searchCrossAppAccessResourceServers(Reference reference, String query, int limit);

    Single<TrustedDomain> create(Domain domain, NewTrustedDomain newTrustedDomain, User principal);

    Single<TrustedDomain> update(Domain domain, String id, UpdateTrustedDomain updateTrustedDomain, User principal);

    @Deprecated
    Single<TrustedDomain> create(Domain domain, NewTrustDomain newTrustDomain, User principal);

    @Deprecated
    Single<TrustedDomain> update(Domain domain, String id, UpdateTrustDomain updateTrustDomain, User principal);

    Completable delete(Domain domain, String id, User principal);
}
